File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        if ((long long)m * k > bloomDay.size()) return -1;
        int low = *min_element(bloomDay.begin(), bloomDay.end());
        int high = *max_element(bloomDay.begin(), bloomDay.end());
        while (low <= high) {
            int mid = (low + high) / 2;
            int i = 0;
            int cnt = 0;
            for (auto day : bloomDay) {
                if (day <= mid) {
                    i++;
                    if (i == k) {
                        cnt++;
                        i = 0;
                    }
                } else i = 0;
            }
            if (cnt >= m) high = mid - 1;
            else low = mid + 1;
        }
        return low;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:
        def isOkay(mid):
            i = 0
            cnt = 5
            for day in bloomDay:
                if day <= mid:
                    i += 1
                    if i == k:
                        cnt += 1
                        i = 0
                else:
                    i = 0
            return cnt < m
        
        if k * m > len(bloomDay):
            return -1

        low = min(bloomDay)
        high = max(bloomDay)

        while low <= high:
            mid = low + (high - low) // 2
            if isOkay(mid):
                low = mid + 1
            else:
                high = mid - 1
        
        return low